File: benchmark/perf.jl
# Script for generating benchmark results in
# https://henry2004y.github.io/Vlasiator.jl/dev/benchmark/

using Vlasiator, PyPlot
using BenchmarkTools

# Turn off GUI for Matplotlib
pygui(false)

files = ["1d_single.vlsv", "bulk.2d.vlsv", "2d_double.vlsv", "2d_AFC.vlsv", "3d_EGI.vlsv"]

# Download test files if not found in the current path
for i in eachindex(files)
   if isfile(files[i])
      @info "Benchmark file $(files[i]) found..."
      continue
   elseif i == 4
      url_base = "https://a3s.fi/swift/v1/AUTH_81f1cd490d494224880ea77e4f98490d/vlasiator-2d-afc/"
      filename = "production_halfres/bulk.0000000.vlsv"
      file_origin = basename(filename)
      url = joinpath(url_base, filename)
      @info "Downloading 1.8G test file..."
      run(`curl -o $(files[i]) $url`)
   elseif i in (1,2,3)
      @info "Downloading test files..."
      run(`curl -o testdata.tar.gz https://raw.githubusercontent.com/henry2004y/vlsv_data/master/testdata.tar.gz`)
      run(`curl -o 1d_single.vlsv https://raw.githubusercontent.com/henry2004y/vlsv_data/master/1d_single.vlsv`)
      run(`curl -o 2d_double.vlsv https://raw.githubusercontent.com/henry2004y/vlsv_data/master/2d_double.vlsv`)
      run(`tar -xzvf testdata.tar.gz`)
   elseif i == 5
      @warn "$(files[i]) is not open-access!"
   end
end

# Define a parent BenchmarkGroup to contain our suite
const suite = BenchmarkGroup()

# Add children groups and tags to our benchmark suite.
suite["load"] = BenchmarkGroup(files)
suite["read"] = BenchmarkGroup(files)
suite["plot"] = BenchmarkGroup(["2d", "3d"])

# Add benchmarks
for (i, file) in enumerate(files)
   !isfile(file) && continue
   suite["load"][file] = @benchmarkable load($file)
   if i == 4
      var = "proton/rho"
      suite["read"][file*"_sorted"] = @benchmarkable meta[$var] setup=(meta=load($file))
      suite["read"][file*"_unsorted"] =
         @benchmarkable readvariable(meta, $var, false) setup=(meta=load($file))
      suite["plot"]["contour_uniform"] =
         @benchmarkable pcolormesh(meta, $var, colorscale=Log) setup=(meta=load($file))
   else
      var = "proton/vg_rho"
      suite["read"][file*"_sorted"] = @benchmarkable meta[$var] setup=(meta=load($file))
      suite["read"][file*"_unsorted"] =
         @benchmarkable readvariable(meta, $var, false) setup=(meta=load($file))
   end
   if i == 5
      var = "proton/vg_rho"
      suite["plot"]["contour_nonuniform"] =
         @benchmarkable pcolormesh(meta, $var, colorscale=Log) setup=(meta=load($file))
   end
end

# If a cache of tuned parameters already exists, use it, otherwise, tune and cache
# the benchmark parameters. Reusing cached parameters is faster and more reliable
# than re-tuning `suite` every time the file is included.
paramspath = joinpath(dirname(@__FILE__), "params.json")

if isfile(paramspath)
   loadparams!(suite, BenchmarkTools.load(paramspath)[1], :evals)
else
   tune!(suite)
   BenchmarkTools.save(paramspath, params(suite))
end

results = run(suite, verbose=true, samples=100, seconds=20)

show(results)

if isfile(files[5])
   println("")
   println("----------")
   meta = load(files[5])
   # mmap
   @time b = readvariable(meta, "fg_b", true, true);
   println("----------")
   # regular
   @time b = readvariable(meta, "fg_b", true, false);
end

File: benchmark/perf.py
# Script for generating benchmark results in
# https://henry2004y.github.io/Vlasiator.jl/dev/benchmark/
#
# If Tex is not installed, LaTeX format output can be disabled via
#    export PTNOLATEX=1

# Set environment variables for Analysator
import os

os.environ["PTNONINTERACTIVE"] = "1"
os.environ["PTBACKEND"] = 'Agg'

import requests, tarfile, timeit, os.path, textwrap
import pytools as pt

files = ['1d_single.vlsv', 'bulk.2d.vlsv', '2d_double.vlsv', '2d_AFC.vlsv', '3d_EGI.vlsv']

for i, file in enumerate(files):
   if os.path.isfile(file):
      print(f"Benchmark file {file} found...")
   elif i == 3:
      url_base = 'https://a3s.fi/swift/v1/AUTH_81f1cd490d494224880ea77e4f98490d/vlasiator-2d-afc/'
      filename = 'production_halfres/bulk.0000000.vlsv'
      r = requests.get(url_base+filename, allow_redirects=True)
      open('2d_AFC.vlsv', 'wb').write(r.content)
   elif i in [0,1,2]:
      r = requests.get('https://raw.githubusercontent.com/henry2004y/vlsv_data/master/testdata.tar.gz', allow_redirects=True)
      open('testdata.tar.gz', 'wb').write(r.content)
      r = requests.get('https://raw.githubusercontent.com/henry2004y/vlsv_data/master/1d_single.vlsv', allow_redirects=True)
      open('1d_single.vlsv', 'wb').write(r.content)
      r = requests.get('https://raw.githubusercontent.com/henry2004y/vlsv_data/master/2d_double.vlsv', allow_redirects=True)
      open('2d_double.vlsv', 'wb').write(r.content)

      file = tarfile.open('testdata.tar.gz')
      file.extractall('./')
      file.close()
   elif i == 4:
      print(f"{file} is not open-access!")

# Number of trails for each benchmark
ntrail = 100

# Setup: importing packages
p0 = """
import pytools as pt
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
matplotlib.use('agg')
"""

for i, file in enumerate(files):
   if not os.path.isfile(file):
      continue
   # Setup: reading metadata
   p2 = textwrap.dedent("""\
   f = pt.vlsvfile.VlsvReader(file)
   cellID = f.read_variable('CellID')
   cell_sorted = np.argsort(cellID)
   """)

   if i != 3:
      # Setup: selecting file and variable
      p1 = p0 + textwrap.dedent(f"""\
      file = '{file}'
      var = 'proton/vg_rho'
      """)
   else:
      # Setup: selecting file and variable
      p1 = p0 + textwrap.dedent(f"""\
      file = '{file}'
      var = 'proton/rho'
      """)

   # Run: reading metadata
   s1 = p2
   t1 = timeit.timeit(stmt=s1, setup=p1, number=ntrail) / ntrail * 1e3
   print(f"{file}:")
   print(f"Reading metadata in {t1:0.4f} ms")
   # Run: obtaining sorted scalar variable
   s2 = textwrap.dedent("""\
   rho = f.read_variable(var)[cell_sorted]
   """)
   t2 = timeit.timeit(stmt=s2, setup=p0+p1+p2, number=ntrail) / ntrail * 1e3
   print(f"{file}:")
   print(f"Reading scalar DCCRG variable in {t2:0.4f} ms")

# Run: plotting density contour on a uniform mesh
s3 = f"""
pt.plot.plot_colormap(filename='{files[3]}', var='rho', draw=1)
"""
t3 = timeit.timeit(stmt=s3, setup=p0, number=5) / 5
print(f"Uniform 2d plotting in {t3:0.4f} s")

if os.path.isfile(files[4]):
   # Run: plotting density slice from a 3D AMR mesh
   s4 = textwrap.dedent(f"""\
   pt.plot.plot_colormap3dslice(filename='{files[4]}', var='proton/vg_rho', draw=1, normal='y')
   """)

   t4 = timeit.timeit(stmt=s4, setup=p0, number=5) / 5
   print(f"AMR slice plotting in {t4:0.4f} s")